File: backend/app/services/conversation_service.py
"""
Conversation lifecycle management.

Handles finding/creating conversations and status transitions.
"""

import uuid

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.models import Conversation, Customer


async def find_or_create_customer(
    db: AsyncSession,
    tenant_id: uuid.UUID,
    phone: str,
) -> Customer:
    """Find customer by phone or create a new one."""
    result = await db.execute(
        select(Customer).where(
            Customer.tenant_id == tenant_id,
            Customer.phone == phone,
        )
    )
    customer = result.scalar_one_or_none()

    if customer is None:
        customer = Customer(tenant_id=tenant_id, phone=phone)
        db.add(customer)
        await db.flush()

    return customer


async def find_or_create_conversation(
    db: AsyncSession,
    tenant_id: uuid.UUID,
    customer_id: uuid.UUID,
) -> Conversation:
    """
    Find an active conversation for this customer or create a new one.

    Active = status is 'new' or 'in_progress'.
    """
    result = await db.execute(
        select(Conversation).where(
            Conversation.tenant_id == tenant_id,
            Conversation.customer_id == customer_id,
            Conversation.status.in_(["new", "in_progress"]),
        )
    )
    conversation = result.scalar_one_or_none()

    if conversation is None:
        conversation = Conversation(
            tenant_id=tenant_id,
            customer_id=customer_id,
            channel="whatsapp",
        )
        db.add(conversation)
        await db.flush()

    return conversation


VALID_TRANSITIONS = {
    "new": {"in_progress", "resolved"},
    "in_progress": {"resolved"},
    "resolved": set(),  # Terminal for MVP
}


async def update_conversation_status(
    db: AsyncSession,
    conversation: Conversation,
    new_status: str,
) -> Conversation:
    """Update conversation status with transition validation."""
    allowed = VALID_TRANSITIONS.get(conversation.status, set())
    if new_status not in allowed:
        raise ValueError(
            f"Cannot transition from '{conversation.status}' to '{new_status}'"
        )

    conversation.status = new_status
    await db.flush()
    return conversation

File: backend-ts/src/services/conversation.service.ts
import type { Conversation } from "@prisma/client";
import { prisma } from "../lib/prisma.js";
import { type Result, ok, fail } from "../lib/result.js";
import { NotFoundError, ValidationError } from "../lib/errors.js";

const VALID_TRANSITIONS: Record<string, string[]> = {
  queued: ["active", "closed"],
  active: ["waiting", "closed"],
  waiting: ["active", "closed"],
  closed: [],
};

/** Strip +, spaces, dashes, parens — keep only digits */
export function normalizePhone(phone: string): string {
  return phone.replace(/[^\d]/g, "");
}

export async function findOrCreateCustomer(
  tenantId: string,
  phone: string,
  name?: string,
  profilePictureUrl?: string,
  tag?: string,
  role?: string,
) {
  // Auto-detect WhatsApp groups by the @g.us suffix
  const isGroup = phone.includes("@g.us");
  const effectiveTag = isGroup ? "GROUP_STAFF" : (tag ?? "GUEST");
  
  // Decide default role if not provided
  let effectiveRole = role;
  if (!effectiveRole) {
    if (isGroup || effectiveTag === "STAFF" || effectiveTag === "GROUP_STAFF") {
      effectiveRole = "staff";
    } else {
      effectiveRole = "lead"; // Por padrão, entra como lead, depois o intent avalia
    }
  }

  const normalized = isGroup ? phone : normalizePhone(phone);

  let customer = await prisma.customer.findUnique({
    where: {
      tenantId_phone: { tenantId, phone: normalized },
    },
  });

  if (!customer) {
    customer = await prisma.customer.create({
      data: {
        tenantId,
        phone: normalized,
        name: name ?? phone,
        profilePictureUrl: profilePictureUrl ?? null,
        tag: effectiveTag,
        role: effectiveRole,
      },
    });
  } else {
    // Update profile picture, tag and/or role if needed
    const updates: Record<string, unknown> = {};
    if (profilePictureUrl && !customer.profilePictureUrl) updates.profilePictureUrl = profilePictureUrl;
    if (isGroup && customer.tag !== "GROUP_STAFF") updates.tag = "GROUP_STAFF";
    if (role && customer.role !== role) updates.role = role;

    if (Object.keys(updates).length > 0) {
      customer = await prisma.customer.update({
        where: { id: customer.id },
        data: updates,
      });
    }
  }

  return customer;
}

export async function findOrCreateConversation(
  tenantId: string,
  customerId: string,
  channel: string,
): Promise<{ conversation: Conversation; isNew: boolean }> {
  // Look for the most recent conversation (including closed ones)
  const existing = await prisma.conversation.findFirst({
    where: {
      tenantId,
      customerId,
      channel,
    },
    orderBy: { updatedAt: "desc" },
  });

  if (existing) {
    // Reopen closed conversations so the same chat is always reused
    if (existing.status === "closed") {
      const reopened = await prisma.conversation.update({
        where: { id: existing.id },
        data: { status: "queued" },
      });
      return { conversation: reopened, isNew: false };
    }
    return { conversation: existing, isNew: false };
  }

  const conversation = await prisma.conversation.create({
    data: {
      tenantId,
      customerId,
      channel,
      status: "queued",
    },
  });

  return { conversation, isNew: true };
}

export async function updateConversationStatus(
  conversationId: string,
  tenantId: string,
  newStatus: string,
): Promise<Result<Conversation, NotFoundError | ValidationError>> {
  const conversation = await prisma.conversation.findFirst({
    where: { id: conversationId, tenantId },
  });

  if (!conversation) {
    return fail(new NotFoundError("Conversation"));
  }

  const allowed = VALID_TRANSITIONS[conversation.status];
  if (!allowed?.includes(newStatus)) {
    return fail(
      new ValidationError(
        `Cannot transition from '${conversation.status}' to '${newStatus}'`,
      ),
    );
  }

  const updated = await prisma.conversation.update({
    where: { id: conversationId },
    data: { status: newStatus },
  });

  return ok(updated);
}
